AVX zip_lo/hi incorrectly uses AVX2 instructions

👋  MWE:

```cpp
#include <xsimd/xsimd.hpp>

using float_v = xsimd::batch<float>;
using uint_v = xsimd::batch<unsigned int>;
using int_v = xsimd::batch<int>;

std::pair<int_v, int_v> interleave(const int_v &a, const int_v &b)
{
    return {xsimd::zip_lo(a, b), xsimd::zip_hi(a, b)};
}
```

When built with `c++ -O2 -mavx` (clang 13 on MSYS; note it's *not* AVX2): 

```
In file included from C:/benchmark-xsimd/main2.cpp:1:
In file included from C:/include/xsimd/xsimd.hpp:29:
In file included from C:/include/xsimd/config/../types/../types/xsimd_batch.hpp:384:
In file included from C:/include/xsimd/config/../types/../arch/xsimd_isa.hpp:48:
C:/include/xsimd/arch/./xsimd_avx.hpp:1184:24: error: always_inline function '_mm256_unpacklo_epi32' requires target feature 'avx2', but would be inlined into function 'zip_lo' that is compiled without support for 'avx2'
                return _mm256_unpacklo_epi32(self, other);
                       ^
C:/include/xsimd/arch/./xsimd_avx.hpp:1154:24: error: always_inline function '_mm256_unpackhi_epi32' requires target feature 'avx2', but would be inlined into function 'zip_hi' that is compiled without support for 'avx2'
                return _mm256_unpackhi_epi32(self, other);
```

Checked in xsimd 8.0.3, and also in commit 315f7c11751bab0771478f48c4197b4cf9b50404.

--- a/include/xsimd/arch/xsimd_avx.hpp
+++ b/include/xsimd/arch/xsimd_avx.hpp
@@ -1146,6 +1146,37 @@ namespace xsimd
         {
             switch (sizeof(T))
             {
+            case 1:
+            case 2:
+            {
+                // extract low word
+                __m128i self_lo = _mm256_extractf128_si256(self, 0);
+                __m128i self_hi = _mm256_extractf128_si256(self, 1);
+                // extract high word
+                __m128i other_lo = _mm256_extractf128_si256(other, 0);
+                __m128i other_hi = _mm256_extractf128_si256(other, 1);
+
+                // interleave
+                __m128i res_lo, res_hi;
+                if (sizeof(T) == 1)
+                {
+                    res_lo = _mm_unpackhi_epi8(self_lo, other_lo);
+                    res_hi = _mm_unpackhi_epi8(self_hi, other_hi);
+                }
+                else
+                {
+                    res_lo = _mm_unpackhi_epi16(self_lo, other_lo);
+                    res_hi = _mm_unpackhi_epi16(self_hi, other_hi);
+                }
+
+                // fuse
+                return _mm256_castps_si256(
+                    _mm256_insertf128_ps(
+                        _mm256_castsi256_ps(_mm256_castsi128_si256(res_lo)),
+                        _mm_castsi128_ps(res_hi),
+                        1));
+            }
+
             case 4:
                 return _mm256_castps_si256(_mm256_unpackhi_ps(_mm256_castsi256_ps(self), _mm256_castsi256_ps(other)));
             case 8:
@@ -1172,6 +1203,36 @@ namespace xsimd
         {
             switch (sizeof(T))
             {
+            case 1:
+            case 2:
+            {
+                // extract low word
+                __m128i self_lo = _mm256_extractf128_si256(self, 0);
+                __m128i self_hi = _mm256_extractf128_si256(self, 1);
+                // extract high word
+                __m128i other_lo = _mm256_extractf128_si256(other, 0);
+                __m128i other_hi = _mm256_extractf128_si256(other, 1);
+
+                // interleave
+                __m128i res_lo, res_hi;
+                if (sizeof(T) == 1)
+                {
+                    res_lo = _mm_unpacklo_epi8(self_lo, other_lo);
+                    res_hi = _mm_unpacklo_epi8(self_hi, other_hi);
+                }
+                else
+                {
+                    res_lo = _mm_unpacklo_epi16(self_lo, other_lo);
+                    res_hi = _mm_unpacklo_epi16(self_hi, other_hi);
+                }
+
+                // fuse
+                return _mm256_castps_si256(
+                    _mm256_insertf128_ps(
+                        _mm256_castsi256_ps(_mm256_castsi128_si256(res_lo)),
+                        _mm_castsi128_ps(res_hi),
+                        1));
+            }
             case 4:
                 return _mm256_castps_si256(_mm256_unpacklo_ps(_mm256_castsi256_ps(self), _mm256_castsi256_ps(other)));
             case 8:
@@ -1192,9 +1253,7 @@ namespace xsimd
         {
             return _mm256_unpacklo_pd(self, other);
         }
-
     }
-
 }
 
 #endif

--- a/include/xsimd/arch/xsimd_avx.hpp
+++ b/include/xsimd/arch/xsimd_avx.hpp
@@ -1146,14 +1146,10 @@ namespace xsimd
         {
             switch (sizeof(T))
             {
-            case 1:
-                return _mm256_unpackhi_epi8(self, other);
-            case 2:
-                return _mm256_unpackhi_epi16(self, other);
             case 4:
-                return _mm256_unpackhi_epi32(self, other);
+                return _mm256_castps_si256(_mm256_unpackhi_ps(_mm256_castsi256_ps(self), _mm256_castsi256_ps(other)));
             case 8:
-                return _mm256_unpackhi_epi64(self, other);
+                return _mm256_castpd_si256(_mm256_unpackhi_pd(_mm256_castsi256_pd(self), _mm256_castsi256_pd(other)));
             default:
                 assert(false && "unsupported arch/op combination");
                 return {};
@@ -1176,19 +1172,16 @@ namespace xsimd
         {
             switch (sizeof(T))
             {
-            case 1:
-                return _mm256_unpacklo_epi8(self, other);
-            case 2:
-                return _mm256_unpacklo_epi16(self, other);
             case 4:
-                return _mm256_unpacklo_epi32(self, other);
+                return _mm256_castps_si256(_mm256_unpacklo_ps(_mm256_castsi256_ps(self), _mm256_castsi256_ps(other)));
             case 8:
-                return _mm256_unpacklo_epi64(self, other);
+                return _mm256_castpd_si256(_mm256_unpacklo_pd(_mm256_castsi256_pd(self), _mm256_castsi256_pd(other)));
             default:
                 assert(false && "unsupported arch/op combination");
                 return {};
             }
         }
+
         template <class A>
         inline batch<float, A> zip_lo(batch<float, A> const& self, batch<float, A> const& other, requires_arch<avx>) noexcept
         {

--- a/include/xsimd/arch/xsimd_avx2.hpp
+++ b/include/xsimd/arch/xsimd_avx2.hpp
@@ -596,6 +596,46 @@ namespace xsimd
                 return sub(self, other, avx {});
             }
         }
+
+        // zip_hi
+        template <class A, class T, class = typename std::enable_if<std::is_integral<T>::value, void>::type>
+        inline batch<T, A> zip_hi(batch<T, A> const& self, batch<T, A> const& other, requires_arch<avx2>) noexcept
+        {
+            switch (sizeof(T))
+            {
+            case 1:
+                return _mm256_unpackhi_epi8(self, other);
+            case 2:
+                return _mm256_unpackhi_epi16(self, other);
+            case 4:
+                return _mm256_unpackhi_epi32(self, other);
+            case 8:
+                return _mm256_unpackhi_epi64(self, other);
+            default:
+                assert(false && "unsupported arch/op combination");
+                return {};
+            }
+        }
+
+        // zip_lo
+        template <class A, class T, class = typename std::enable_if<std::is_integral<T>::value, void>::type>
+        inline batch<T, A> zip_lo(batch<T, A> const& self, batch<T, A> const& other, requires_arch<avx2>) noexcept
+        {
+            switch (sizeof(T))
+            {
+            case 1:
+                return _mm256_unpacklo_epi8(self, other);
+            case 2:
+                return _mm256_unpacklo_epi16(self, other);
+            case 4:
+                return _mm256_unpacklo_epi32(self, other);
+            case 8:
+                return _mm256_unpacklo_epi64(self, other);
+            default:
+                assert(false && "unsupported arch/op combination");
+                return {};
+            }
+        }
     }
 
 }

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -170,7 +170,7 @@ set(XSIMD_TESTS
     test_power.cpp
     test_rounding.cpp
     test_select.cpp
-    test_shuffle_128.cpp
+    test_shuffle.cpp
     test_traits.cpp
     test_trigonometric.cpp
     test_utils.hpp

--- a/test/test_shuffle.cpp
+++ b/test/test_shuffle.cpp
@@ -17,7 +17,7 @@
 namespace
 {
     template <typename T, std::size_t N>
-    struct init_shuffle_128_base
+    struct init_shuffle_base
     {
         using shuffle_vector_type = std::array<T, N>;
         shuffle_vector_type lhs_in, rhs_in, exp_lo, exp_hi;
@@ -27,23 +27,32 @@ namespace
             std::vector<shuffle_vector_type> vects;
             vects.reserve(4);
 
+            constexpr size_t K = 128 / (sizeof(T) * 8);
+            constexpr size_t P = N / K;
+
             /* Generate input data: lhs, rhs */
-            for (size_t i = 0; i < N; ++i)
+            for (size_t p = 0; p < P; ++p)
             {
-                lhs_in[i] = 2 * i + 1;
-                rhs_in[i] = 2 * i + 2;
+                for (size_t i = 0; i < K; ++i)
+                {
+                    lhs_in[i + p * K] = 2 * i + 1;
+                    rhs_in[i + p * K] = 2 * i + 2;
+                }
             }
             vects.push_back(std::move(lhs_in));
             vects.push_back(std::move(rhs_in));
 
             /* Expected shuffle data */
-            for (size_t i = 0, j = 0; i < N / 2; ++i, j = j + 2)
+            for (size_t p = 0; p < P; ++p)
             {
-                exp_lo[j] = lhs_in[i];
-                exp_hi[j] = lhs_in[i + N / 2];
+                for (size_t i = 0, j = 0; i < K / 2; ++i, j = j + 2)
+                {
+                    exp_lo[j + p * K] = lhs_in[i];
+                    exp_hi[j + p * K] = lhs_in[i + K / 2];
 
-                exp_lo[j + 1] = rhs_in[i];
-                exp_hi[j + 1] = rhs_in[i + N / 2];
+                    exp_lo[j + 1 + p * K] = rhs_in[i];
+                    exp_hi[j + 1 + p * K] = rhs_in[i + K / 2];
+                }
             }
             vects.push_back(std::move(exp_lo));
             vects.push_back(std::move(exp_hi));
@@ -54,21 +63,21 @@ namespace
 }
 
 template <class B>
-class shuffle_128_test : public testing::Test
+class shuffle_test : public testing::Test
 {
 protected:
     using batch_type = B;
     using value_type = typename B::value_type;
     static constexpr size_t size = B::size;
 
-    shuffle_128_test()
+    shuffle_test()
     {
         std::cout << "shuffle-128 test" << std::endl;
     }
 
-    void shuffle_128_low_high()
+    void shuffle_low_high()
     {
-        init_shuffle_128_base<value_type, size> shuffle_base;
+        init_shuffle_base<value_type, size> shuffle_base;
         auto shuffle_base_vecs = shuffle_base.create_vectors();
         auto v_lhs = shuffle_base_vecs[0];
         auto v_rhs = shuffle_base_vecs[1];
@@ -80,22 +89,18 @@ protected:
         B b_exp_lo = B::load_unaligned(v_exp_lo.data());
         B b_exp_hi = B::load_unaligned(v_exp_hi.data());
 
-        /* Only Test 128bit */
-        if ((sizeof(value_type) * size) == 16)
-        {
-            B b_res_lo = xsimd::zip_lo(b_lhs, b_rhs);
-            EXPECT_BATCH_EQ(b_res_lo, b_exp_lo) << print_function_name("shuffle-128 low test");
+        B b_res_lo = xsimd::zip_lo(b_lhs, b_rhs);
+        EXPECT_BATCH_EQ(b_res_lo, b_exp_lo) << print_function_name("zip low test");
 
-            B b_res_hi = xsimd::zip_hi(b_lhs, b_rhs);
-            EXPECT_BATCH_EQ(b_res_hi, b_exp_hi) << print_function_name("shuffle-128 high test");
-        }
+        B b_res_hi = xsimd::zip_hi(b_lhs, b_rhs);
+        EXPECT_BATCH_EQ(b_res_hi, b_exp_hi) << print_function_name("zip high test");
     }
 };
 
-TYPED_TEST_SUITE(shuffle_128_test, batch_types, simd_test_names);
+TYPED_TEST_SUITE(shuffle_test, batch_types, simd_test_names);
 
-TYPED_TEST(shuffle_128_test, shuffle_128_low_high)
+TYPED_TEST(shuffle_test, shuffle_low_high)
 {
-    this->shuffle_128_low_high();
+    this->shuffle_low_high();
 }
 #endif

